scaper/core.py

```diff
@@ -1698,38 +1698,42 @@ def _generate_audio(self, audio_path, ann, reverb=None,
 
         with _set_temp_logging_level(temp_logging_level):
 
-            # Array for storing all tmp files (one for every event)
-            tmpfiles = []
-            with _close_temp_files(tmpfiles):
-                isolated_events_audio_path = []
+            # List for storing all generated audio (one array for every event)
+            event_audio_arrays = []
+            isolated_events_audio_path = []
+            duration_in_samples = int(self.duration * self.sr)
 
-                role_counter = {'background': 0, 'foreground': 0}
+            role_counter = {'background': 0, 'foreground': 0}
 
+            # TODO: This is for a cleaner diff, remove it after CR.
+            tmpfiles = []
+            with _close_temp_files(tmpfiles):
                 for i, e in enumerate(ann.data):
                     if e.value['role'] == 'background':
-                        # Concatenate background if necessary. Right now we
-                        # always concatenate the background at least once,
-                        # since the pysox combiner raises an error if you try
-                        # to call build using an input_file_list with less than
-                        # 2 elements. In the future if the combiner is updated
-                        # to accept a list of length 1, then the max(..., 2)
-                        # statement can be removed from the calculation of
-                        # ntiles.
-                        source_duration = (
-                            soundfile.info(e.value['source_file']).duration)
+                        # Concatenate background if necessary.
+                        source_duration = soundfile.info(e.value['source_file']).duration
                         ntiles = int(
-                            max(self.duration // source_duration + 1, 2))
+                            max(self.duration // source_duration + 1, 1))
 
-                        # Create combiner
-                        cmb = sox.Combiner()
+                        # Create transformer
+                        tfm = sox.Transformer()
                         # Ensure consistent sampling rate and channels
-                        cmb.convert(samplerate=self.sr,
-                                    n_channels=self.n_channels,
-                                    bitdepth=None)
+                        # Need both a convert operation (to do the conversion),
+                        # and set_output_format (to have sox interpret the output
+                        # correctly).
+                        tfm.convert(
+                            samplerate=self.sr,
+                            n_channels=self.n_channels,
+                            bitdepth=None
+                        )
+                        tfm.set_output_format(
+                            rate=self.sr,
+                            channels=self.n_channels
+                        )
```

Comment:
Why is this needed? Is it because there's no output filename now?

Reply:
Yeah, we need to tell `sox` explicitly what the output format is so it can interpret what it's putting onto `stdout` correctly.